File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> ans;
    int n;
    void solve(string &s,int open,int closed)
    {
        if(open==n && closed==n)
        {
            ans.push_back(s);
            return;
        }

        if(open<n)
        {
            s.push_back('(');
            solve(s,open+1,closed);
            s.pop_back();
        }

        if(closed<open)
        {
            s.push_back(')');
            solve(s,open,closed+1);
            s.pop_back();
        }
    }
    
    vector<string> generateParenthesis(int N) 
    {
        string s="";
        n=N;
        solve(s,0,0);
        return ans;
    }
};

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    private void generate(int open, int closed, StringBuilder sb, List<String> res){
        if(closed == 0){
            res.add(sb.toString());
            return;
        }
        
        if(open < 0){
            return;
        }
        
        sb.append('(');
        generate(open - 1, closed, sb, res);
        sb.deleteCharAt(sb.length() - 1);
        
        if(open < closed){
            sb.append(')');
            generate(open, closed - 1, sb, res);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
    
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        generate(n, n, new StringBuilder(), res);
        return res;
    }
}